### PR TITLE
Powerups

### PR DESCRIPTION
**Pull request by acheronfail**
Here's an initial proposal to #77.

- Powerups spawn approx. every 30 seconds
- There are 3 types
    - Triple shot
    - Shield
    - Time slow
- They last for 10 seconds
- If not picked up within 10 seconds, they disappear

Items to resolve before merging:

- [x] Sometimes the player dies even when a shield is active (I think the collisions fn is called multiple times?)

### CODE REVIEW

**Comment by aochagavia** [comment]
Awesome! I am looking forward to review this, hopefully before the end of the week

File: src/controllers/collisions.rs [hunk 1]
```diff
@@ -62,20 +74,77 @@ impl CollisionsController {
...
-            // Play player_destroyed sound
-            let _ = resources.player_destroyed_sound.play();
-            return true;
+        let mut player_died = false;
+        let player = &mut state.world.player;
+
+        if !player.is_dead && state.world.enemies.iter().any(|enemy| player.collides_with(enemy)) {
+            // Remove shield powerup from player, also killing any enemies within close range
+            let mut had_shield = false;
+            if let Some(powerup) = player.powerup {
```
Comment (aochagavia) [inline src/controllers/collisions.rs]: I think you could use `if let Some(PowerupKind::Shield) = player.powerup` here. That way you get rid of the if below.

File: src/controllers/time.rs [hunk 2]
```diff
@@ -46,49 +50,95 @@ impl TimeController {
...
+        #[cfg(feature = "debug")]
+        let dt = dt * 0.1;
+
         self.current_time += dt;
         state.difficulty += dt / 100.0;
 
-        self.update_player(dt, actions, state);
+        // Check if we have the "TimeSlow" powerup
+        let mut time_slow = false;
+        if let Some(powerup) = state.world.player.powerup {
```
Comment (aochagavia) [inline src/controllers/time.rs]: See my previous comment on `if let`

File: src/controllers/time.rs [hunk 3]
```diff
@@ -46,49 +50,95 @@ impl TimeController {
...
+                        Bullet::new(Vector::new(pos, dir + f64::consts::PI / 6.0)),
+                    ]);
+                }
+            }
+
+            // If there was no powerup, shoot normally
+            if !did_have_powerup {
+                let vector = Vector::new(state.world.player.front(), state.world.player.direction());
+                state.world.bullets.push(Bullet::new(vector));
+            }
```
Comment (aochagavia) [inline src/controllers/time.rs]: I would rewrite the code above in a different style:

```rust
match state.world.player.powerup {
    Some(PowerupKind::TipleShot) {
        // The code you already have inside the `if let`
    }
    _ => {
        // The code you alreade have inside `if !did_have_powerup`
    }
}
```

File: src/controllers/collisions.rs [hunk 4]
```diff
@@ -62,20 +74,77 @@ impl CollisionsController {
...
+    }
+
+    fn remove_surrounding_enemies(enemies: &mut Vec<Enemy>, particles: &mut Vec<Particle>, point: Point) {
+        util::fast_retain(enemies, |enemy| {
+            let enemy_pos = enemy.position();
+            if enemy_pos.intersect_circle(&point, PLAYER_GRACE_AREA) {
+                util::make_explosion(particles, &enemy_pos, 10);
+                return false;
+            } else {
+                return true;
```
Comment (aochagavia) [inline src/controllers/collisions.rs]: I would prefer plain `true` and `false` here, instead of using `return`

**Comment by aochagavia** [comment]
Great job! Just tested the result. I didn't observe the issue of the shield not working, though. Did you figure out how to solve it or was it just luck?

File: src/controllers/collisions.rs (hunk 1 above)
Comment (acheronfail) [reply src/controllers/collisions.rs]: DId not know you could unwrap enums like that - fantastic!

**Comment by acheronfail** [comment]
No I didn't - change `PowerupKind::random` to only choose Shields and then set the spawn rate of Powerups to something low - run into groups of enemies and you'll see that you also explode as well. From what I can tell the `CollisionsController::handle_player_collisions` is called twice rapidly (first time removes the shield, second kills the player). 

I'll do some further investigating and see if I've missed some sort of edge case or something.

**Comment by acheronfail** [comment]
Just tried it as above: even with a shield the player definitely does die when there's a large group of enemies - also, strangely, not all the enemies in the area are removed as well when the shield is destroyed (iirc `CollisionsController:remove_surrounding_enemies` should remove them all).

File: src/controllers/time.rs [hunk 5]
```diff
@@ -78,10 +78,8 @@ impl TimeController {
 
         // Check if we have the "TimeSlow" powerup
         let mut time_slow = false;
-        if let Some(powerup) = state.world.player.powerup {
-            if powerup == PowerupKind::TimeSlow {
-                time_slow = true;
-            }
+        if let Some(PowerupKind::TimeSlow) = state.world.player.powerup {
+            time_slow = true;
         }
```
Comment (aochagavia) [inline src/controllers/time.rs]: What about `let time_slow = state.world.player.powerup == Some(PowerupKind::TimeSlow)`?

Comment (acheronfail) [reply src/controllers/time.rs]: Ha! It seems I continue to miss the obvious... 😅

**Comment by aochagavia** [comment]
I think I found a bug in the implementation of `fast_retain`... I'll do some testing and post the results here

**Comment by aochagavia** [comment]
Confirmed. This is a bug in `fast_retain`, which sometimes didn't remove elements that should have been removed. Just pushed a fix to the master branch. Everything should work fine after rebasing (hopefully)

**Comment by aochagavia** [comment]
Should have written some unit tests for that... 🤦‍♂️

**Comment by acheronfail** [comment]
Haha don't worry, we've all made those mistakes!
I merged in the changed in the latest commit, I'll test it and double check that it's alright.

**Comment by acheronfail** [comment]
Works like a charm! 👍 🎉

**Comment by aochagavia** [comment]
Great, merging!

**Comment by aochagavia** [comment]
Thanks!

**Pull request merged**